File: lib/Snake.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plane_1 = require("./plane");
const Point_1 = require("./Point");
const Body_1 = require("./snake/Body");
const Segment_1 = require("./snake/Segment");
const default_body = new Body_1.default([new Segment_1.default(new Point_1.default(0, 0), new Point_1.default(2, 0))]);
class Snake {
    static withDirection(start, dir) {
        return new Snake(new Body_1.default([
            new Segment_1.default(start, start.add(dir.delta)),
        ]));
    }
    constructor(body = default_body) {
        if (body.isValid() === false) {
            throw new Error('segments dont connect');
        }
        this.body = body;
    }
    head() {
        return this.body.head();
    }
    tail() {
        return this.body.tail();
    }
    up() {
        return this.moveTo(plane_1.default.up);
    }
    right() {
        return this.moveTo(plane_1.default.right);
    }
    down() {
        return this.moveTo(plane_1.default.down);
    }
    left() {
        return this.moveTo(plane_1.default.left);
    }
    direction() {
        return this.head().direction(plane_1.default);
    }
    grow() {
        return new Snake(this.body.prolong());
    }
    forward() {
        const dir = this.direction();
        if (dir === undefined) {
            throw new Error('no direction');
        }
        return this.moveTo(dir);
    }
    moveTo(new_dir) {
        const dir = this.direction();
        if (dir === undefined) {
            return Snake.withDirection(this.head().start, plane_1.default.up);
        }
        else if (dir.delta.additiveInverseOf(new_dir.delta) === false) {
            // cant move backwards
            // add to head
            const prolonged_body = this.body.turnTo(new_dir.delta);
            // shorten tail
            const new_body = prolonged_body.shorten();
            return new Snake(new_body.sanitize());
        }
        else {
            return this;
        }
    }
}
exports.default = Snake;

File: src/Snake.py
from src.Direction import Direction, Orientation
from src.plane import plane
from src.Point import Point, ZERO
from src.snake.Body import Body
from src.snake.Segment import Segment

default_body = Body([Segment(Point(0, 0), Point(2, 0))])

class InvalidMovement(Exception):
  pass

class Snake(object):
  @staticmethod
  def withDirection(start, dir):
    return Snake.withLength(start, dir, 1)

  @staticmethod
  def withLength(start, dir, length):
    return Snake(Body([
      Segment(start, start.add(Point(dir.delta.x * length, dir.delta.y * length))),
    ]))

  def __init__(self, body = default_body):
    if body.isValid() is False:
      raise Exception('segments dont connect')

    self.body = body

  def mouth(self):
    return self.head().end

  def head(self):
    return self.body.head()

  def tail(self):
    return self.body.tail()

  def up(self):
    return self._moveTo(plane.up)

  def right(self):
    return self._moveTo(plane.right)

  def down(self):
    return self._moveTo(plane.down)

  def left(self):
    return self._moveTo(plane.left)

  def direction(self):
    return self.head().direction(plane)

  def grow(self):
    return Snake(self.body.prolong())

  def forward(self):
    dir = self.direction()

    if dir is None:
      raise Exception('no direction')

    return self._moveTo(dir)

  # private
  def _moveTo(self, new_dir):
    dir = self.direction()

    if dir is None:
      return Snake.withDirection(self.head().start, plane.up)
    elif dir.delta.additiveInverseOf(new_dir.delta) is False:
      # cant move backwards

      # add to head
      prolonged_body = self.body.turnTo(new_dir.delta)

      # shorten tail
      new_body = prolonged_body.shorten()

      return Snake(new_body.sanitize())
    else:
      raise InvalidMovement('no movement backwards allowed')
